asgineer/utils.py

```diff
@@ -79,75 +79,82 @@ async def some_handler(request):
-    for key, val in assets.items():
+    for path, body in assets.items():
+        # Get lowercase path
+        lpath = path.lower()
         # Get binary body
-        if isinstance(val, bytes):
-            bbody = val
-        elif isinstance(val, str):
-            bbody = val.encode()
+        if isinstance(body, bytes):
+            bbody = body
+        elif isinstance(body, str):
+            bbody = body.encode()
         else:
-            raise ValueError("Asset bodies must be str or bytes.")
-        # Store hash
-        hashes[key] = hashlib.sha256(bbody).hexdigest()
-        # Store zipped version if it makes sense
+            raise ValueError("Asset bodies must be bytes or str.")
+        # Store etag
+        etags[lpath] = hashlib.sha256(bbody).hexdigest()
+        # Store unzipped body
+        unzipped[lpath] = bbody
+        # Store zipped body if it makes sense
         if len(bbody) >= min_compress_size:
-            if not key.endswith(VIDEO_EXTENSIONS):
+            if not lpath.endswith(VIDEO_EXTENSIONS):
                 bbody_zipped = gzip.compress(bbody)
                 if len(bbody_zipped) < 0.90 * len(bbody):
-                    zipped[key] = bbody_zipped
-        # Store mimetype
-        ctype, enc = mimetypes.guess_type(key)
+                    zipped[lpath] = bbody_zipped
+        # Store ctype
+        ctype, _ = mimetypes.guess_type(lpath)
         if ctype:
-            ctypes[key] = ctype
+            ctypes[lpath] = ctype
         else:
-            ctypes[key] = guess_content_type_from_body(val)
+            ctypes[lpath] = guess_content_type_from_body(body)
 
     async def asset_handler(request, path=None):
         if request.method not in ("GET", "HEAD"):
             return 405, {}, "Method not allowed"
+
         if path is None:
             path = request.path.lstrip("/")
         path = path.lower()
 
-        # Exit early on 404
-        if path not in assets:
-            return 404, {}, "404 not found"
-
-        content_etag = hashes.get(path, None)
-        request_etag = request.headers.get("if-none-match", None)
-        headers = {
-            "etag": content_etag,
-            "cache-control": f"public, must-revalidate, max-age={max_age:d}",
-        }
+        if path not in unzipped:
+            return 404, {}, "File not found"
 
-        # If client already has the exact asset, send confirmation now
-        if request_etag and request_etag == content_etag:
-            return 304, headers, b""
+        assert path in etags
+        assert path in ctypes
 
-        # Set content type
-        if path in ctypes:
-            headers["content-type"] = ctypes[path]
+        status = 200
+        headers = {}
+        headers["cache-control"] = f"public, must-revalidate, max-age={max_age:d}"
+        headers["content-length"] = str(len(unzipped[path]))
+        headers["content-type"] = ctypes[path]
+        headers["etag"] = f"\"{etags[path]}\""
+        body = unzipped[path]
```

Comment:
```suggestion
        status = 200
        body = unzipped[path]
        headers = {}
        headers["cache-control"] = f"public, must-revalidate, max-age={max_age:d}"
        headers["content-length"] = str(len(body))
        headers["content-type"] = ctypes[path]
        headers["etag"] = f"\"{etags[path]}\""
```

Reply:
mmm, the diff looks like a lot has changed, but I only moved the `body = ...` up, and then used `body` where content-length is set. Basically the same change as the other suggestion.